src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -557,74 +691,16 @@ ITypeSymbol IArrayCreationExpression.ElementType
-            {
-                BoundArrayInitialization initializer = this.InitializerOpt;
-                if (initializer != null)
-                {
-                    return MakeInitializer(initializer);
-                }
-
-                return null; 
-            }
-        }
-
-        private static System.Runtime.CompilerServices.ConditionalWeakTable<BoundArrayInitialization, IArrayInitializer> ArrayInitializerMappings = new System.Runtime.CompilerServices.ConditionalWeakTable<BoundArrayInitialization, IArrayInitializer>();
-
-        IArrayInitializer MakeInitializer(BoundArrayInitialization initializer)
-        {
-            return ArrayInitializerMappings.GetValue(
-                initializer,
-                (arrayInitializer) =>
-                {
-                    ArrayBuilder<IArrayInitializer> dimension = ArrayBuilder<IArrayInitializer>.GetInstance(arrayInitializer.Initializers.Length);
-                    for (int index = 0; index < arrayInitializer.Initializers.Length; index++)
-                    {
-                        BoundExpression elementInitializer = arrayInitializer.Initializers[index];
-                        BoundArrayInitialization elementArray = elementInitializer as BoundArrayInitialization;
-                        dimension[index] = elementArray != null ? MakeInitializer(elementArray) : new ElementInitializer(elementInitializer);
-                    }
-
-                    return new DimensionInitializer(dimension.ToImmutableAndFree());
-                });
-        }
+        IArrayInitializer IArrayCreationExpression.Initializer => (IArrayInitializer)(this.InitializerOpt);
 
         protected override OperationKind ExpressionKind => OperationKind.ArrayCreationExpression;
-
-        class ElementInitializer : IExpressionArrayInitializer
-        {
-            private readonly BoundExpression _element;
-
-            public ElementInitializer(BoundExpression element)
-            {
-                _element = element;
-            }
-
-            public IExpression ElementValue => _element;
-
-            public ArrayInitializerKind ArrayClass => ArrayInitializerKind.Expression;
-        }
-
-        class DimensionInitializer : IDimensionArrayInitializer
-        {
-            private readonly ImmutableArray<IArrayInitializer> _dimension;
-
-            public DimensionInitializer(ImmutableArray<IArrayInitializer> dimension)
-            {
-                _dimension = dimension;
-            }
-
-            public ImmutableArray<IArrayInitializer> ElementValues => _dimension;
-
-            public ArrayInitializerKind ArrayClass => ArrayInitializerKind.Dimension;
-        }
     }
 
-    partial class BoundArrayInitialization
+    partial class BoundArrayInitialization : IArrayInitializer
     {
-        protected override OperationKind ExpressionKind => OperationKind.None;
+        public ImmutableArray<IExpression> ElementValues => this.Initializers.As<IExpression>();
```

Comment:
Consider explicit implementation instead if all callers will be calling with expressions of type `IArrayInitializer`.